contrib/mypy/src/python/pants/contrib/mypy/tasks/mypy_task.py

```diff
@@ -0,0 +1,132 @@
+from pants.backend.python.targets.python_binary import PythonBinary
+from pants.backend.python.targets.python_library import PythonLibrary
+from pants.backend.python.targets.python_target import PythonTarget
+from pants.backend.python.targets.python_tests import PythonTests
+from pants.backend.python.tasks.python_task import PythonTask
+from pants.base.build_environment import get_buildroot
+from pants.base.exceptions import TaskError
+from pants.base.workunit import WorkUnit, WorkUnitLabel
+from pants.util.contextutil import temporary_dir
+from pants.util.process_handler import subprocess
+from pex.pex import PEX
+from pex.pex_builder import PEXBuilder
+from pex.pex_info import PexInfo
+
+
+class MypyTask(PythonTask):
+  """Invoke the mypy static type analyzer for Python."""
+
+  _PYTHON_SOURCE_EXTENSION = '.py'
+
+  @classmethod
+  def register_options(cls, register):
+    register('--mypy-version', default='0.550', help='The version of mypy to use')
+
+  @classmethod
+  def supports_passthru_args(cls):
+    return True
+
+  def py3_interpreter(self):
+    interpreters = self._interpreter_cache.setup(filters=['CPython>=3'])
+    return min(interpreters) if interpreters else None
+
+  @staticmethod
+  def is_non_synthetic_python_target(target):
+    return not target.is_synthetic and isinstance(target, (PythonLibrary, PythonBinary, PythonTests))
+
+  @staticmethod
+  def is_python_target(target):
+    return isinstance(target, (PythonTarget,))
+
+  def _calculate_python_sources(self, targets):
+    """Generate a set of source files from the given targets."""
+    python_eval_targets = filter(self.is_non_synthetic_python_target, targets)
+    sources = set()
+    for target in python_eval_targets:
+      sources.update(
+        source for source in target.sources_relative_to_buildroot()
+        if os.path.splitext(source)[1] == self._PYTHON_SOURCE_EXTENSION
+      )
+    return list(sources)
+
+  def _calculate_python_package_roots(self, targets):
+    """Calculate the set of directories at the top of each subtree of Python sources."""
+
+    # Collect the set of directories in which there are Python sources (whether part of
+    # the target roots or transitive dependencies.)
+    directories = set()
+    for target in self.context.targets(self.is_python_target):
+      if not target.has_sources(self._PYTHON_SOURCE_EXTENSION):
+        continue
+
+      for source in target.sources_relative_to_buildroot():
+        directories.add(os.path.dirname(source))
+
+    # For each directory, find the first parent directory without an __init__.py file.
+    # This should be the package root.
+    roots = set()
+    for directory in directories:
+      while directory:
+        if not os.path.exists(os.path.join(get_buildroot(), directory, '__init__.py')):
+          break
+        directory = os.path.dirname(directory)
+
+      if directory:
+        roots.add(directory)
+
+    import pdb ; pdb.set_trace()
```

Comment:
?

Reply:
Debugging leftover. Pushed too early. Made better progress locally. Will push when the integration test is working.